Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -409,9 +444,14 @@ extension NIOAsyncWriter {
         @inlinable
         /* fileprivate */ internal init(
```

Comment:
There are still two `deinit`s in `NIOAsyncWriter` that haven't been fixed here: https://github.com/apple/swift-nio/pull/2582/files#diff-7914f7c4968a26adfb4f6be410d7a1c12a72a35ed7bfd1c3a62f08e08f4b37f1R178-R376

Reply:
Oh I actually forgot to call the new method in one of those. The other one was already correct. Both will unconditionally call through to the state machine and we make the decision what happens in the state machine.

Reply:
Can we not flip this around to make this reviewable and very clear: Unless the user tells us to, we're not gonna do anything (but assertions) in `deinit`. If `deinit` seems to do a thing (like calling something else) then you lose local reasoning and have to follow all possible trails to check if that code is actually `deinit`-safe Swift code. If it's just

```
deinit {
    if self.yoloMode {
       self.stateMachine.doStuff()
    }
}
```

then it's super clear that nothing bad is going to happen

Reply:
`deinit` shouldn't drive state machines, it shouldn't do anything if at all possible.

Reply:
I disagree with this. I prefer to have all reasoning in the state machine and only have a single place that takes decisions. Furthermore, we must involve the state machine anyhow because we need to understand if we are already finished. I would prefer to keep it like I implemented here.

Reply:
I do understand what you're saying but `deinit` is a very special case. Code that's in there runs under special rules (must not publish `self` or else UB or crash). If you want anybody to understand this code you have to have it in `deinit`.

Imagine this deinit

```
deinit {
    self.stateMachine.finish()
}
```

or 

```
deinit {
    self.finish()
}
```

looks nice and simple. If a future PR changes `func finish()` in the state machine or on `self`, then the reviewer of that PR will have _no clue_ that this code may be called from `deinit`. Normally, the rules are simple: No code is called from deinit that isn't. If we bend that rule it's really important that this is visible in every diff that touches it. Hiding it in the state machine might have dire consequences that can be ~impossible to fix. If Swift had a mandatory `@mayBeCalledFromDeinit` attribute this were different but we don't

Reply:
That's why we are not calling `finish` in the state machine but specific methods called `writerDeinitialized` or `sinkDenitialized`. This makes it very explicit to understand where this call comes from. IMO that's better than calling out to the state machine in deinit to check for a specific state and then make the decision what to do outside of the state machine. That's just blurs the line where state logic happens.